presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/LogFilter.java

```diff
@@ -0,0 +1,286 @@
+        {
+            requireNonNull(groupName);
+            if (filter != null) {
+                namedGroupFilters.put(groupName, filter);
+            }
+            return this;
+        }
+
+        /**
+         * Add a filter to apply to a capturing group in
+         * the regular expression for log file lines
+         */
+        public Builder addGroupFilter(Integer groupNumber, Predicate<String> filter)
+        {
+            requireNonNull(groupNumber);
+            if (filter != null) {
+                numberedGroupFilters.put(groupNumber, filter);
+            }
+            return this;
+        }
+
+        /**
+         * Use this string to separate lines in the same entry
+         */
+        public Builder setLineSeparator(String separator)
+        {
+            requireNonNull(separator);
+            lineSeparator = separator;
+            return this;
+        }
+
+        /**
+         * Set the maximum number of entries to keep when filtering.
+         *
+         * A null capacity results in all entries being kept.
+         */
+        public Builder setCapacity(Integer capacity)
+        {
+            maxEntries = capacity;
+            return this;
+        }
+
+        /**
+         * When there is a maximum capacity, keep the first entries
+         * instead of the last entries
+         */
+        public Builder keepFirst()
+        {
+            keepFirst = true;
+            return this;
+        }
+    }
+
+    /**
+     * Class representing a collection of log entries
+     */
+    private final class LogEntries
+    {
+        private Deque<StringJoiner> entries;
+        private Boolean keptLastLine;
+
+        private LogEntries()
+        {
+            entries = (maxEntries == null)
+                    ? new ArrayDeque<>()
+                    : new ArrayDeque<>(maxEntries);
+        }
+
+        Stream<StringJoiner> streamEntries()
+        {
+            return entries.stream();
+        }
+
+        /**
+         * Add a line from a log file to a deque of log entries
+         *
+         * If the string does not match {@link #logPattern}, it will be
+         * appended to the previous entry. If there is no previous entry, a new
+         * entry will be created only if entries are not being filtered.
+         *
+         * If {@link #maxEntries} is non-null and the deque contains that many
+         * entries, the oldest will be removed before the new entry is added.
+         */
+        void addLine(String line)
+        {
+            Matcher matcher = logPattern.matcher(line);
+            if (matcher.matches()) {
+                /* Apply all filters */
+                boolean passing = true;
+                for (ImmutableMap.Entry<String, Predicate<String>> f
+                        : namedGroupFilters.entrySet()) {
+                    passing = passing && f.getValue().test(matcher.group(f.getKey()));
```

Comment:
Second part does take a little more time, but I guess performance at this point does not matter that much, up to you.